### ISSUE TITLE
[Immix.cpp] Thread sanitizer data race warnings when using haxe threads

### ISSUE THREAD

**Issue by haxiomic** [HaxeFoundation/hxcpp#882]
Testing with hxcpp 4.0.52 and haxe 4.0.5 on macOS 10.14

I was trying to ensure my hxcpp threading code was sound by using `-fsanitize=thread` but I always get data-race warnings in the hxcpp GC code. I reduced the project and found the warnings occur in a minimal project with only a few self-contained threads performing allocations

Here's the haxe code
```haxe
class Main {

    static function main() {
        trace('Starting up some threads that perform allocations');

        haxe.EntryPoint.addThread(() -> {
            while(true) {
                var array = [];
                for (i in 0...100) {
                    array.push(i);
                }
                Sys.sleep(1/1000);
            }
        });

        haxe.EntryPoint.addThread(() -> {
            while(true) {
                var array = [];
                for (i in 0...100) {
                    array.push(i);
                }
                Sys.sleep(1/1000);
            }
        });
    }

}
```

An example warning is
```
WARNING: ThreadSanitizer: data race (pid=80636)
  Write of size 4 at 0x000106e15030 by thread T1 (mutexes: write M163):
    #0 GlobalAllocator::MarkAll(bool) Immix.cpp:4505 (Main-debug:x86_64+0x10015867f)
    #1 GlobalAllocator::Collect(bool, bool, bool) Immix.cpp:4726 (Main-debug:x86_64+0x100142431)
    #2 GlobalAllocator::GetFreeBlock(int, hx::ImmixAllocator*) Immix.cpp:3455 (Main-debug:x86_64+0x100156845)
    #3 LocalAllocator::CallAlloc(int, unsigned int) Immix.cpp:5911 (Main-debug:x86_64+0x100155f63)
    #4 hx::ImmixAllocator::alloc(hx::ImmixAllocator*, unsigned long, bool, char const*) GC.h:426 (Main-debug:x86_64+0x100002bdc)
    #5 hx::Object::operator new(unsigned long, bool, char const*) Object.h:170 (Main-debug:x86_64+0x1000027fa)
    #6 Array_obj<int>::__new(int, int) Array.h:1085 (Main-debug:x86_64+0x100005e18)
    #7 Main_obj::main()::_hx_Closure_0::_hx_run() Main.cpp:47 (Main-debug:x86_64+0x100005ce9)
    #8 Main_obj::main()::_hx_Closure_0::__run() Main.cpp:59 (Main-debug:x86_64+0x100005b93)
    #9 hxThreadFunc(void*) Thread.cpp:249 (Main-debug:x86_64+0x100133176)

  Previous read of size 4 at 0x000106e15030 by thread T2:
    #0 LocalAllocator::CallAlloc(int, unsigned int) Immix.cpp:5875 (Main-debug:x86_64+0x100155cb6)
    #1 hx::InternalNew(int, bool) Immix.cpp:6225 (Main-debug:x86_64+0x100141966)
    #2 hx::ArrayBase::Realloc(int) const Array.cpp:116 (Main-debug:x86_64+0x100033178)
    #3 hx::ArrayBase::EnsureSize(int) const Array.h:325 (Main-debug:x86_64+0x100007ed7)
    #4 Array_obj<int>::push(int) Array.h:620 (Main-debug:x86_64+0x100005ff6)
    #5 Main_obj::main()::_hx_Closure_1::_hx_run() Main.cpp:71 (Main-debug:x86_64+0x100012db9)
    #6 Main_obj::main()::_hx_Closure_1::__run() Main.cpp:77 (Main-debug:x86_64+0x100012c83)
    #7 hxThreadFunc(void*) Thread.cpp:249 (Main-debug:x86_64+0x100133176)

  As if synchronized via sleep:
    #0 nanosleep <null>:1061488 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x29703)
    #1 _hx_std_sys_sleep(double) Sys.cpp:133 (Main-debug:x86_64+0x1001b64e3)
    #2 Main_obj::main()::_hx_Closure_0::_hx_run() Main.cpp:56 (Main-debug:x86_64+0x100005d81)
    #3 Main_obj::main()::_hx_Closure_0::__run() Main.cpp:59 (Main-debug:x86_64+0x100005b93)
    #4 hxThreadFunc(void*) Thread.cpp:249 (Main-debug:x86_64+0x100133176)

  Location is global 'hx::gMarkID' at 0x000106e15030 (Main-debug+0x0001001eb030)

  Mutex M163 (0x7b1400001818) created at:
    #0 pthread_mutex_init <null>:1061248 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2b973)
    #1 HxMutex::HxMutex() Thread.h:197 (Main-debug:x86_64+0x100135513)
    #2 HxMutex::HxMutex() Thread.h:193 (Main-debug:x86_64+0x1001325e8)
    #3 GlobalAllocator::AddLocal(LocalAllocator*) Immix.cpp:3050 (Main-debug:x86_64+0x100155893)
    #4 LocalAllocator::LocalAllocator(int*) Immix.cpp:5463 (Main-debug:x86_64+0x1001556b3)
    #5 LocalAllocator::LocalAllocator(int*) Immix.cpp:5454 (Main-debug:x86_64+0x100141148)
    #6 hx::InitAlloc() Immix.cpp:6141 (Main-debug:x86_64+0x100140ed3)
    #7 hx::SetTopOfStack(int*, bool) Immix.cpp:6166 (Main-debug:x86_64+0x10014126a)
    #8 main HxcppMain.h:88 (Main-debug:x86_64+0x100032519)

  Thread T1 (tid=1020628, running) created by main thread at:
    #0 pthread_create <null>:1061440 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2a99d)
    #1 HxCreateDetachedThread(void* (*)(void*), void*) Thread.h:228 (Main-debug:x86_64+0x1001337c2)
    #2 __hxcpp_thread_create(Dynamic) Thread.cpp:278 (Main-debug:x86_64+0x100133459)
    #3 Main_obj::main() Main.cpp:96 (Main-debug:x86_64+0x1000045fa)
    #4 __hxcpp_main __main__.cpp:25 (Main-debug:x86_64+0x10003275a)
    #5 main HxcppMain.h:93 (Main-debug:x86_64+0x100032537)

  Thread T2 (tid=1020629, running) created by main thread at:
    #0 pthread_create <null>:1061440 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2a99d)
    #1 HxCreateDetachedThread(void* (*)(void*), void*) Thread.h:228 (Main-debug:x86_64+0x1001337c2)
    #2 __hxcpp_thread_create(Dynamic) Thread.cpp:278 (Main-debug:x86_64+0x100133459)
    #3 Main_obj::main() Main.cpp:97 (Main-debug:x86_64+0x10000466f)
    #4 __hxcpp_main __main__.cpp:25 (Main-debug:x86_64+0x10003275a)
    #5 main HxcppMain.h:93 (Main-debug:x86_64+0x100032537)

SUMMARY: ThreadSanitizer: data race Immix.cpp:4505 in GlobalAllocator::MarkAll(bool)
```

Do you think these are false positives / nothing to worry about or likely to be pointing to bugs?

Here's a repro project:
[hxcpp-tsan.zip](https://github.com/HaxeFoundation/hxcpp/files/4177501/hxcpp-tsan.zip)

**Comment by haxiomic** [comment]
Another example warning:
```
WARNING: ThreadSanitizer: data race (pid=82557)
  Write of size 4 at 0x7b480000069c by thread T2 (mutexes: write M163):
    #0 hx::QuickVec<BlockDataInfo*>::push(BlockDataInfo* const&) QuickVec.h:32 (Main-debug:x86_64+0x10013b503)
    #1 GlobalAllocator::AllocMoreBlocks(bool&, bool) Immix.cpp:3385 (Main-debug:x86_64+0x10013aef0)
    #2 GlobalAllocator::GetFreeBlock(int, hx::ImmixAllocator*) Immix.cpp:3448 (Main-debug:x86_64+0x1001450aa)
    #3 LocalAllocator::CallAlloc(int, unsigned int) Immix.cpp:5911 (Main-debug:x86_64+0x100144853)
    #4 hx::InternalRealloc(void*, int, bool) Immix.cpp:6318 (Main-debug:x86_64+0x100132261)
    #5 hx::ArrayBase::Realloc(int) const Array.cpp:109 (Main-debug:x86_64+0x100019ee4)
    #6 hx::ArrayBase::EnsureSize(int) const Array.h:325 (Main-debug:x86_64+0x100008107)
    #7 Array_obj<int>::push(int) Array.h:620 (Main-debug:x86_64+0x100006226)
    #8 Main_obj::main()::_hx_Closure_1::_hx_run() Main.cpp:64 (Main-debug:x86_64+0x100013039)
    #9 Main_obj::main()::_hx_Closure_1::__run() Main.cpp:70 (Main-debug:x86_64+0x100012f03)
    #10 hxThreadFunc(void*) Thread.cpp:249 (Main-debug:x86_64+0x100121a66)

  Previous read of size 4 at 0x7b480000069c by thread T1:
    #0 hx::QuickVec<BlockDataInfo*>::size() const QuickVec.h:120 (Main-debug:x86_64+0x100139e76)
    #1 GlobalAllocator::GetNextFree(int) Immix.cpp:3254 (Main-debug:x86_64+0x100145530)
    #2 GlobalAllocator::GetFreeBlock(int, hx::ImmixAllocator*) Immix.cpp:3424 (Main-debug:x86_64+0x100144f6b)
    #3 LocalAllocator::CallAlloc(int, unsigned int) Immix.cpp:5911 (Main-debug:x86_64+0x100144853)
    #4 hx::ImmixAllocator::alloc(hx::ImmixAllocator*, unsigned long, bool, char const*) GC.h:426 (Main-debug:x86_64+0x10000352c)
    #5 hx::Object::operator new(unsigned long, bool, char const*) Object.h:170 (Main-debug:x86_64+0x10000314a)
    #6 Array_obj<int>::__new(int, int) Array.h:1085 (Main-debug:x86_64+0x100006048)
    #7 Main_obj::main()::_hx_Closure_0::_hx_run() Main.cpp:40 (Main-debug:x86_64+0x100005ec9)
    #8 Main_obj::main()::_hx_Closure_0::__run() Main.cpp:52 (Main-debug:x86_64+0x100005d73)
    #9 hxThreadFunc(void*) Thread.cpp:249 (Main-debug:x86_64+0x100121a66)

  As if synchronized via sleep:
    #0 nanosleep <null>:1598608 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x29703)
    #1 _hx_std_sys_sleep(double) Sys.cpp:133 (Main-debug:x86_64+0x1001a4dd3)
    #2 Main_obj::main()::_hx_Closure_1::_hx_run() Main.cpp:67 (Main-debug:x86_64+0x100013071)
    #3 Main_obj::main()::_hx_Closure_1::__run() Main.cpp:70 (Main-debug:x86_64+0x100012f03)
    #4 hxThreadFunc(void*) Thread.cpp:249 (Main-debug:x86_64+0x100121a66)

  Location is heap block of size 328 at 0x7b4800000600 allocated by main thread:
    #0 operator new(unsigned long) <null>:1598512 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x70c5b)
    #1 hx::InitAlloc() Immix.cpp:6132 (Main-debug:x86_64+0x10012f64d)
    #2 hx::SetTopOfStack(int*, bool) Immix.cpp:6166 (Main-debug:x86_64+0x10012fb5a)
    #3 main HxcppMain.h:88 (Main-debug:x86_64+0x100019239)

  Mutex M163 (0x7b1400001818) created at:
    #0 pthread_mutex_init <null>:1598368 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2b973)
    #1 HxMutex::HxMutex() Thread.h:197 (Main-debug:x86_64+0x100123e03)
    #2 HxMutex::HxMutex() Thread.h:193 (Main-debug:x86_64+0x100120ed8)
    #3 GlobalAllocator::AddLocal(LocalAllocator*) Immix.cpp:3050 (Main-debug:x86_64+0x100144183)
    #4 LocalAllocator::LocalAllocator(int*) Immix.cpp:5463 (Main-debug:x86_64+0x100143fa3)
    #5 LocalAllocator::LocalAllocator(int*) Immix.cpp:5454 (Main-debug:x86_64+0x10012fa38)
    #6 hx::InitAlloc() Immix.cpp:6141 (Main-debug:x86_64+0x10012f7c3)
    #7 hx::SetTopOfStack(int*, bool) Immix.cpp:6166 (Main-debug:x86_64+0x10012fb5a)
    #8 main HxcppMain.h:88 (Main-debug:x86_64+0x100019239)

  Thread T2 (tid=1044192, running) created by main thread at:
    #0 pthread_create <null>:1598560 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2a99d)
    #1 HxCreateDetachedThread(void* (*)(void*), void*) Thread.h:228 (Main-debug:x86_64+0x1001220b2)
    #2 __hxcpp_thread_create(Dynamic) Thread.cpp:278 (Main-debug:x86_64+0x100121d49)
    #3 Main_obj::main() Main.cpp:75 (Main-debug:x86_64+0x100004fbf)
    #4 __hxcpp_main __main__.cpp:22 (Main-debug:x86_64+0x10001947a)
    #5 main HxcppMain.h:93 (Main-debug:x86_64+0x100019257)

  Thread T1 (tid=1044191, running) created by main thread at:
    #0 pthread_create <null>:1598560 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2a99d)
    #1 HxCreateDetachedThread(void* (*)(void*), void*) Thread.h:228 (Main-debug:x86_64+0x1001220b2)
    #2 __hxcpp_thread_create(Dynamic) Thread.cpp:278 (Main-debug:x86_64+0x100121d49)
    #3 Main_obj::main() Main.cpp:74 (Main-debug:x86_64+0x100004f4a)
    #4 __hxcpp_main __main__.cpp:22 (Main-debug:x86_64+0x10001947a)
    #5 main HxcppMain.h:93 (Main-debug:x86_64+0x100019257)

SUMMARY: ThreadSanitizer: data race QuickVec.h:32 in hx::QuickVec<BlockDataInfo*>::push(BlockDataInfo* const&)
```

**Comment by hughsando** [comment]
I know this is an old thread, but FYI i have looked over the code and there are quite a few warnings, but I think they are now ok.  I moved a few things around to make it clearer and I think there may have been something subtle with register capture code order which should now be fixed.
I started to add a few conditionals to block the core from the sanitizer, but there are quite a few - like non-haxe code that changes the stack while the collection is going on.  And a lot of the allocation code is designed to be "lock free" (like the GetNextFree shown here)  which falsely triggers the checks.
It might be worthwhile to comment each of the  cases, but quite a bit of work, so not sure.
